### ISSUE TITLE
Video propragation IOU mask score

### ISSUE THREAD

**Issue by 2nzi** [facebookresearch/sam2#666]
Hi everyone,

I am looking for the iou mask score during the 'propagate_in_video' . I saw a score return for image prediction but i don't find any score for video (except the 'object_score_logits' wich i don't understand the goal, should i apply a sigmoid function on it ?).

**Comment by heyoeyo** [comment]
The `object_score_logits` are something different from IoU, it's like a score telling you how likely it is that the prediction is valid (eg. low scores or below 0 imply the tracking is broken).

The IoU isn't provided as-is so you'd have to modify the code to be able to get it out of the `propagate_in_video` function. Internally the IoUs are generated as one of the outputs of the [sam_mask_decoder](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/modeling/sam2_base.py#L347C13-L347C17), and used to [auto-select the mask](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/modeling/sam2_base.py#L381-L388) prediction. So you'd have to return the `ious` result through the chain of calls all the way to the `propagate_in_video` function output. It's a pretty deep chain of functions:

[_run_single_frame_inference](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/sam2_video_predictor.py#L603) > [track_step](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/sam2_video_predictor.py#L762) > [_track_step](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/modeling/sam2_base.py#L835) > [_forward_sam_heads](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/modeling/sam2_base.py#L779) > [sam_mask_decoder](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/modeling/sam2_base.py#L347)

So the idea would be to modify the last function to return the `ious`, then works backwards, modifying functions as needed to account for the new `ious` value, until you can get it to the `propagate_in_video` output.

Edit: Actually, it looks like the `ious` value is passed back to the [track_step](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/modeling/sam2_base.py#L852) call (but not used), so you'd only need to modify from there. However, you may still want to alter the later functions to return all the masks (if you're wanting to pick something different based on iou).

**Comment by zaiquanyang** [comment]
> ious

Hi, according to your explanation here and https://github.com/facebookresearch/segment-anything/issues/495, I print the output as the IoU prediction of each video frame. My understanding should be correct?https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/modeling/sam2_base.py#L852 



If that is right, I find the IoU score prediction on my below case is a little interesting. 
The initial few frames successfully predicted masks also with high IoU-scores (>0.9), while the subsequent frames failed to obtain masks and yielded extremely low IoU-scores (<0.001).  So is there a correlation between the presence of mask predictions and the IoU score? Could you provide some insights for this?

<img width="805" height="369" alt="Image" src="https://github.com/user-attachments/assets/d39c8700-61c2-4a92-843e-948736d70e35" />

**Comment by heyoeyo** [comment]
> So is there a correlation between the presence of mask predictions and the IoU score?

Yes, I think there generally should be a correlation. The IoU is the model's guess about how closely it's mask prediction overlaps with whatever the 'correct' prediction would have been. If the model loses track of the object, then it's prediction isn't going to match the correct prediction much/at all, so the IoU should be small.

It's like if someone gave you 5 seconds to draw the mask on a frame, and then asked on a scale of 0-100%, how accurate do you think your mask was? Even though you don't have time to figure out the 'perfect' mask in 5 seconds, you can (quickly) approximate how close your attempt is to being perfect. That's sort of what the IoU prediction is doing. If you couldn't find the object in 5 seconds (e.g. [where's waldo](https://stackoverflow.com/questions/8849869/how-do-i-find-wally-with-python)), then you wouldn't be able to draw a mask and you'd guess ~0%.

That all being said, the `object_score_logits` tends to be a more reliable indicator of when the tracking loses the object (it becomes a negative number), in case that's what you're trying to detect.
